txindex upgrade

ptschip opened the pull request:
Move txindex to it's own folder.  Migrate legacy data if necessary.

Commit:
[index] Create new TxIndex class.

The TxIndex will be responsible for building the transaction index
concurrently with the main validation thread by implementing
ValidationInterface. This does not process blocks concurrently yet.

Commit:
[index] TxIndex initial sync thread.

TxIndex starts up a background thread to get in sync with the block
index before blocks are processed through the ValidationInterface.

Commit:
[index] Allow TxIndex sync thread to be interrupted.

Commit:
[index] TxIndex method to wait until caught up.

In order to preserve getrawtransaction RPC behavior, there needs to be
a way for a thread to ensure the transaction index is in sync with the
current state of the blockchain.

Commit:
[init] Initialize and start TxIndex in init code.

Commit:
[db] Create separate database for txindex.

The new TxIndexDB class will be used by a future commit in this
change set.

Commit:
[db] Migration for txindex data to new, separate database.

ptschip committed:
Move txindex logic from  GetTransaction() to TxIndex::FindTx()

ptschip committed:
Remove unused functions

Remove unused variables

ptschip committed:
Commit the new txns to the txindex database in ConnectBlock()

WriteBestBlock in BlockConnected()

This is to ensure that we've already synced the txindex and we
don't end up writing a better block than we updated the txindex
to.  Generally it wouldn't be an issue but if someone restarts
bitcoind before the txindex is fully up to date then they could
end up with an invalid best block and end up skipping many
blocks.

Also, remove SetBestChain as it wasn't being used and is not
necessary in BU.

ptschip committed:
Fix error in params

ptschip committed:
Remove unnecessary check for fTxIndex in init.cpp

ptschip committed:
Tidy up the progress messages for the upgrade

ptschip committed:
Don't allow rpc until the txindex is fully synced

For some of the rpc's that make calls to GetTransation() we want
to pring out a message that the txindex is not yet available
and to try again later once it's fully synced.

ptschip committed:
Adjust parameter naming to be more in alignement with BU's codebase

ptschip committed:
Set the fReindex flag when starting txindex

Remove unnecessary section in blockconnected()

ptschip committed:
Do not writebest block if we are shutting down but have not written the txindex data

ptschip committed:
Remove the Read and Write flags for txindex

We don't need these anymore.  It doesn't matter if we turn txindex off
after we had it on and visa versa. All we need to know is if we
currently have turned txindex on via the config settings.

ptschip committed:
Change BlockUntilSynced to chain to IsSynced().

This is much clearer as to what this function is doing. It's not
blocking at all but rather returning a bool indicating the synced
status.

ptschip committed:
Add python tests

ptschip committed:
Tidy up cs_main locking

ptschip commented:
Ready for testing.  I've done quite a few upgrades, stops/starts etc..seems to work fine.  The port was not a clean one and had to make quite a few little mods to get it to work with ours.

ptschip committed:
Fix up a bad rebase

Some other PR was merged which had somewhat major conflicts with
the setting of cache values. This commit fixes that up but it
would be better, in the future, to pass a struct around rather
than adding new parameters which are passed to the functions each
time we need a new one.

gandrewstone commented on src/main.cpp:
```diff
@@ -216,6 +217,7 @@ void UnregisterNodeSignals(CNodeSignals &nodeSignals)
 CBlockIndex *FindForkInGlobalIndex(const CChain &chain, const CBlockLocator &locator)
 {
     // Find the first block the caller has in the main chain
+    LOCK(cs_main); // for chain
```
chain is parameter so we don't know what critsec protects it.  The caller needs to grab the appropriate lock (which is going to be cs_main for chainActive) before calling this function.  It looks like this is already happening in 2 of the 4 places where this function is called.

ptschip replied:
good point , done.

gandrewstone commented on src/init.cpp:
```diff
@@ -1277,6 +1283,12 @@ bool AppInit2(Config &config, thread_group &threadGroup)
+    if (GetBoolArg("-txindex", DEFAULT_TXINDEX))
+    {
+        auto txindex_db = std::make_unique<TxIndexDB>(nTxIndexCache, false, fReindex);
+        g_txindex = std::make_unique<TxIndex>(std::move(txindex_db));
```
I don't love the wordiness of std::make_unique and std::move.  I see these functions as useful when 99% of the time you want a resource allocation to follow the stack, but rarely you need to give ownership of it to some other code.  

But here we are creating a pointer txindex_db just to hand ownership off to someone else.  Why not just pass the pointer and let TxIndex wrap the reference into a unique_ptr?:

```
    if (GetBoolArg("-txindex", DEFAULT_TXINDEX))
    {
        auto txindex_db = new TxIndexDB(nTxIndexCache, false, fReindex);
        g_txindex = std::make_unique<TxIndex>(txindex_db);
        g_txindex->Start();
    }

txindex.cpp:
TxIndex::TxIndex(TxIndexDB* _db) : db(_db), fSynced(false), pbestindex(nullptr) {}

txindex.h:
    /// Constructs the TxIndex, which becomes available to be queried.  TxIndex takes ownership of db
    explicit TxIndex(TxIndexDB* db);
```

This uses a lot less C++ complexity and machinery, likely creates a lot less code (especially in debug builds), and doesn't pass stuff through several temporaries.

ptschip replied:
fixed.

ptschip replied:
I didn't write this code but actually I think this is correct.  Both the g_txindex and the txindexdb are unique pointers...the txindexdb being the db inside the g_txindex object.  There is no way around it unless we entirely abandon the smart pointer for the txindexdb and then we'll just be back to using a regular pointer with a delete on destruction without the benefit of the smart pointer.  I would prefer to keep the smart pointer because if we in the future have other smart pointers (which seems to be the trend within the codebase) within the db object then they'll get destructed in the right order..i've painfully been through this with graphene and is why I switched all the sub-objects there to smart pointers.

gandrewstone replied:
I still have 2 smart pointers (TxIndex member and g_txindex) .  The original code had 4.  The 2 you are not seeing is the temporary created in the if GetBoolArg scope and the temporary parameter of the TxIndex ctor.

ptschip replied:
ok you're right,  done.

gandrewstone reviewed (changes requested):
looking at this code, it seems possible to dynamically turn txindexing on and off without a restart, since the block processing thread runs essentially independently.  If so, let's make this a tweak so that we can do it.  Changing the tweak value would start and stop the txindexing thread...

ptschip committed:
The caller should lock cs_main for FindForkInGlobalIndex()

chain is passed to the function for which cs_main is needed so
the caller should really be locking cs_main before calling.

ptschip committed:
Simplify the intialization of g_txindex and txindex_db
